modules/ml/src/svmsgd.cpp

```diff
@@ -0,0 +1,595 @@
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#include "precomp.hpp"
+#include "limits"
+
+#include <iostream>
+
+using std::cout;
+using std::endl;
+
+/****************************************************************************************\
+*                        Stochastic Gradient Descent SVM Classifier                      *
+\****************************************************************************************/
+
+namespace cv
+{
+namespace ml
+{
+
+class SVMSGDImpl : public SVMSGD
+{
+
+public:
+    SVMSGDImpl();
+
+    virtual ~SVMSGDImpl() {}
+
+    virtual bool train(const Ptr<TrainData>& data, int);
+
+    virtual float predict( InputArray samples, OutputArray results=noArray(), int flags = 0 ) const;
+
+    virtual bool isClassifier() const;
+
+    virtual bool isTrained() const;
+
+    virtual void clear();
+
+    virtual void write(FileStorage &fs) const;
+
+    virtual void read(const FileNode &fn);
+
+    virtual Mat getWeights(){ return weights_; }
+
+    virtual float getShift(){ return shift_; }
+
+    virtual int getVarCount() const { return weights_.cols; }
+
+    virtual String getDefaultName() const {return "opencv_ml_svmsgd";}
+
+    virtual void setOptimalParameters(int svmsgdType = ASGD, int marginType = SOFT_MARGIN);
+
+    virtual int getSvmsgdType() const;
+
+    virtual void setSvmsgdType(int svmsgdType);
+
+    virtual int getMarginType() const;
+
+    virtual void setMarginType(int marginType);
+
+    CV_IMPL_PROPERTY(float, Lambda, params.lambda)
+    CV_IMPL_PROPERTY(float, Gamma0, params.gamma0)
+    CV_IMPL_PROPERTY(float, C, params.c)
+    CV_IMPL_PROPERTY_S(cv::TermCriteria, TermCriteria, params.termCrit)
+
+private:
+    void updateWeights(InputArray sample, bool isFirstClass, float gamma, Mat &weights);
+
+    std::pair<bool,bool> areClassesEmpty(Mat responses);
+
+    void writeParams( FileStorage &fs ) const;
+
+    void readParams( const FileNode &fn );
+
+    static inline bool isFirstClass(float val) { return val > 0; }
+
+    static void normalizeSamples(Mat &matrix, Mat &average, float &multiplier);
+
+    float calcShift(InputArray _samples, InputArray _responses) const;
+
+    static void makeExtendedTrainSamples(const Mat &trainSamples, Mat &extendedTrainSamples, Mat &average, float &multiplier);
+
+
+
+    // Vector with SVM weights
+    Mat weights_;
+    float shift_;
+
+    // Parameters for learning
+    struct SVMSGDParams
+    {
+        float lambda;                             //regularization
```

Comment:
I guess the parameters can, in fact, have better names than just some Greek letters =)